File: Conceitos-Fundamentais/Alocação-Dinâmica/alocacao-dinamica.js
// Exemplo de alocação dinâmica em JavaScript

//Alocação Dinâmica é visível em estruturas de dados como arrays e objetos, onde o tamanho pode ser alterado durante a execução do programa.

// Criando um array de números dinamicamente
let numeros = [];  // Inicializando um array vazio

// Alocando elementos dinamicamente
numeros.push(10);
numeros.push(20);
numeros.push(30);

console.log("Array de números:", numeros);

// O array pode ser alterado a qualquer momento, adicionando ou removendo elementos
numeros.push(40);
console.log("Array após adicionar mais um número:", numeros);

// E a memória alocada pode ser ajustada conforme necessário
numeros.pop();  // Remove o último elemento
console.log("Array após remover um número:", numeros);

File: Conceitos-Fundamentais/Alocação-Dinâmica/alocacao-dinamica.c
#include <stdio.h>
#include <stdlib.h>

int main() {
    // Alocando dinamicamente um vetor de inteiros com 5 elementos
    int *vetor = (int *)malloc(5 * sizeof(int));  // aloca memória para 5 inteiros

    if (vetor == NULL) {
        printf("Falha na alocação de memória.\n");
        return 1;  // Sai do programa caso a alocação falhe
    }

    // Inicializando o vetor com valores
    for (int i = 0; i < 5; i++) {
        vetor[i] = i * 10;
    }

    // Exibindo os valores do vetor
    printf("Vetor alocado dinamicamente: ");
    for (int i = 0; i < 5; i++) {
        printf("%d ", vetor[i]);
    }
    printf("\n");

    // Liberando a memória alocada dinamicamente
    free(vetor);

    return 0;
}
